### ISSUE TITLE
Connection for few client closing repeatedly

### ISSUE THREAD

**Issue by daadu** [emqx/emqx#3228]
#### Environment

- OS: Ubuntu 18.04.1
- Erlang/OTP: R22/10.5.2
- EMQ: v4.0-rc.3

#### Description
2 clients out of 10 in my broker are reconnecting continuously. I could find following in the log.

Is the issue with broker or client?

Logs for device 1
```
2020-01-31T11:34:32.512102175Z node.name=d0fdc9b37869@172.17.0.2
2020-01-31T11:34:32.757754240Z EMQX_LOADED_PLUGINS=emqx_recon,emqx_retainer,emqx_management,emqx_dashboard,emqx_auth_pgsql
2020-01-31T11:34:50.259901241Z EMQ X Broker v4.0-rc.3 is started successfully!
2020-01-31T11:34:56.469965624Z ['2020-01-31T11:34:56Z']:emqx start
2020-01-31T11:35:01.263726045Z 2020-01-31 11:35:00.974 [error] <<"esp32_D9AAD0">>@180.151.108.70:61993   crasher:
2020-01-31T11:35:01.263845651Z     initial call: emqx_connection:init/4
2020-01-31T11:35:01.263937748Z     pid: <0.1813.0>
2020-01-31T11:35:01.264034641Z     registered_name: []
2020-01-31T11:35:01.264161312Z     exception exit: {badarg,
2020-01-31T11:35:01.264243263Z                         [{ets,lookup_element,
2020-01-31T11:35:01.264335495Z                              [emqx_metrics,'packets.publish.inuse',4],
2020-01-31T11:35:01.264429779Z                              []},
2020-01-31T11:35:01.264522941Z                          {emqx_metrics,update_counter,2,
2020-01-31T11:35:01.264612112Z                              [{file,
2020-01-31T11:35:01.264704975Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_metrics.erl"},
2020-01-31T11:35:01.264804587Z                               {line,292}]},
2020-01-31T11:35:01.264898890Z                          {emqx_channel,do_publish,3,
2020-01-31T11:35:01.264987932Z                              [{file,
2020-01-31T11:35:01.265076979Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_channel.erl"},
2020-01-31T11:35:01.265173982Z                               {line,413}]},
2020-01-31T11:35:01.265263044Z                          {emqx_connection,with_channel,3,
2020-01-31T11:35:01.265352030Z                              [{file,
2020-01-31T11:35:01.265443167Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:01.265535754Z                               {line,535}]},
2020-01-31T11:35:01.265694145Z                          {emqx_connection,process_msg,3,
2020-01-31T11:35:01.265783023Z                              [{file,
2020-01-31T11:35:01.265869008Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:01.265961666Z                               {line,282}]},
2020-01-31T11:35:01.266068163Z                          {proc_lib,wake_up,3,
2020-01-31T11:35:01.266167425Z                              [{file,"proc_lib.erl"},{line,259}]}]}
2020-01-31T11:35:01.266257625Z       in function  emqx_connection:terminate/2 (/emqx_rel/_checkouts/emqx/src/emqx_connection.erl, line 422)
2020-01-31T11:35:01.266349026Z     ancestors: [<0.1716.0>,<0.1715.0>,esockd_sup,<0.1339.0>]
2020-01-31T11:35:01.266439805Z     message_queue_len: 0
2020-01-31T11:35:01.266528988Z     messages: []
2020-01-31T11:35:01.266830698Z     links: [<0.1716.0>]
2020-01-31T11:35:01.266940886Z     dictionary: [{guid,{1580470500974435,221714801755925,4}},
2020-01-31T11:35:01.267032185Z                   {send_pkt,1},
2020-01-31T11:35:01.267124372Z                   {'$logger_metadata$',
2020-01-31T11:35:01.267364398Z                       #{clientid => <<"esp32_D9AAD0">>,
2020-01-31T11:35:01.267372228Z                         peername => "180.151.108.70:61993"}},
2020-01-31T11:35:01.267375577Z                   {acl_cache_size,4},
2020-01-31T11:35:01.267378431Z                   {incoming_pubs,2},
2020-01-31T11:35:01.267381299Z                   {acl_keys_q,
2020-01-31T11:35:01.267384100Z                       {[{subscribe,<<"43/28/rpc">>},
2020-01-31T11:35:01.267387852Z                         {subscribe,<<"43/28/rpc/#">>},
2020-01-31T11:35:01.267390962Z                         {subscribe,<<"43/28/rw">>}],
2020-01-31T11:35:01.267393964Z                        [{publish,<<"43/28/hello">>}]}},
2020-01-31T11:35:01.267397129Z                   {{subscribe,<<"43/28/rpc/#">>},{allow,1580470500972}},
2020-01-31T11:35:01.267400230Z                   {incoming_bytes,324},
2020-01-31T11:35:01.267403179Z                   {{subscribe,<<"43/28/rw">>},{allow,1580470500971}},
2020-01-31T11:35:01.267662584Z                   {{publish,<<"43/28/hello">>},{allow,1580470500969}},
2020-01-31T11:35:01.267669271Z                   {outgoing_bytes,4},
2020-01-31T11:35:01.267672227Z                   {rand_seed,
2020-01-31T11:35:01.267675075Z                       {#{bits => 58,jump => #Fun<rand.13.8986388>,
2020-01-31T11:35:01.267678263Z                          next => #Fun<rand.10.8986388>,type => exsss,
2020-01-31T11:35:01.267681466Z                          uniform => #Fun<rand.11.8986388>,
2020-01-31T11:35:01.267684455Z                          uniform_n => #Fun<rand.12.8986388>},
2020-01-31T11:35:01.267687429Z                        [209913954079579216|96186559135385903]}},
2020-01-31T11:35:01.267690512Z                   {{subscribe,<<"43/28/rpc">>},{allow,1580470500974}},
2020-01-31T11:35:01.267704327Z                   {recv_msg,2},
2020-01-31T11:35:01.267707255Z                   {recv_pkt,6}]
2020-01-31T11:35:01.267710028Z     trap_exit: false
2020-01-31T11:35:01.267712863Z     status: running
2020-01-31T11:35:01.267715619Z     heap_size: 1598
2020-01-31T11:35:01.267718407Z     stack_size: 27
2020-01-31T11:35:01.267723137Z     reductions: 22422
2020-01-31T11:35:01.267726070Z   neighbours:
2020-01-31T11:35:01.267731769Z 2020-01-31 11:35:00.976 [error]     supervisor: 'esockd_connection_sup - <0.1716.0>'
2020-01-31T11:35:01.267736461Z     errorContext: connection_crashed
2020-01-31T11:35:01.267835149Z     reason: {badarg,
2020-01-31T11:35:01.267850853Z                 [{ets,lookup_element,
2020-01-31T11:35:01.267854466Z                      [emqx_metrics,'packets.publish.inuse',4],
2020-01-31T11:35:01.267857364Z                      []},
2020-01-31T11:35:01.267860169Z                  {emqx_metrics,update_counter,2,
2020-01-31T11:35:01.267863102Z                      [{file,"/emqx_rel/_checkouts/emqx/src/emqx_metrics.erl"},
2020-01-31T11:35:01.267866270Z                       {line,292}]},
2020-01-31T11:35:01.267869144Z                  {emqx_channel,do_publish,3,
2020-01-31T11:35:01.267872052Z                      [{file,"/emqx_rel/_checkouts/emqx/src/emqx_channel.erl"},
2020-01-31T11:35:01.267875018Z                       {line,413}]},
2020-01-31T11:35:01.267877741Z                  {emqx_connection,with_channel,3,
2020-01-31T11:35:01.267880700Z                      [{file,
2020-01-31T11:35:01.267883576Z                           "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:01.267886671Z                       {line,535}]},
2020-01-31T11:35:01.267889442Z                  {emqx_connection,process_msg,3,
2020-01-31T11:35:01.267892228Z                      [{file,
2020-01-31T11:35:01.267895024Z                           "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:01.267898123Z                       {line,282}]},
2020-01-31T11:35:01.267900857Z                  {proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]}
2020-01-31T11:35:01.267903842Z     offender: [{pid,<0.1813.0>},
2020-01-31T11:35:01.273996305Z                {name,connection},
2020-01-31T11:35:01.274016505Z                {mfargs,
2020-01-31T11:35:01.274020577Z                    {emqx_connection,start_link,
2020-01-31T11:35:01.274023688Z                        [[{deflate_options,[]},
2020-01-31T11:35:01.274026717Z                          {max_conn_rate,500},
2020-01-31T11:35:01.274041501Z                          {active_n,100},
2020-01-31T11:35:01.274044434Z                          {zone,external}]]}}]
```

Similarly with device 2:
```
2020-01-31T11:35:06.268738576Z 2020-01-31 11:35:05.435 [error] <<"esp32_D9AB18">>@49.207.53.3:2896   crasher:
2020-01-31T11:35:06.268852633Z     initial call: emqx_connection:init/4
2020-01-31T11:35:06.268943579Z     pid: <0.1830.0>
2020-01-31T11:35:06.269032809Z     registered_name: []
2020-01-31T11:35:06.269120468Z     exception exit: {badarg,
2020-01-31T11:35:06.269216040Z                         [{ets,lookup_element,
2020-01-31T11:35:06.269306542Z                              [emqx_metrics,'packets.publish.inuse',4],
2020-01-31T11:35:06.269394911Z                              []},
2020-01-31T11:35:06.269485043Z                          {emqx_metrics,update_counter,2,
2020-01-31T11:35:06.269573513Z                              [{file,
2020-01-31T11:35:06.269664124Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_metrics.erl"},
2020-01-31T11:35:06.269754650Z                               {line,292}]},
2020-01-31T11:35:06.269845108Z                          {emqx_channel,do_publish,3,
2020-01-31T11:35:06.269954312Z                              [{file,
2020-01-31T11:35:06.270033880Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_channel.erl"},
2020-01-31T11:35:06.270121920Z                               {line,413}]},
2020-01-31T11:35:06.270244114Z                          {emqx_connection,with_channel,3,
2020-01-31T11:35:06.270325656Z                              [{file,
2020-01-31T11:35:06.270407011Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:06.270487001Z                               {line,535}]},
2020-01-31T11:35:06.270627157Z                          {emqx_connection,process_msg,3,
2020-01-31T11:35:06.270716879Z                              [{file,
2020-01-31T11:35:06.270799032Z                                   "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:06.270878950Z                               {line,282}]},
2020-01-31T11:35:06.270958293Z                          {proc_lib,wake_up,3,
2020-01-31T11:35:06.271037978Z                              [{file,"proc_lib.erl"},{line,259}]}]}
2020-01-31T11:35:06.271118266Z       in function  emqx_connection:terminate/2 (/emqx_rel/_checkouts/emqx/src/emqx_connection.erl, line 422)
2020-01-31T11:35:06.271206855Z     ancestors: [<0.1716.0>,<0.1715.0>,esockd_sup,<0.1339.0>]
2020-01-31T11:35:06.273950236Z     message_queue_len: 0
2020-01-31T11:35:06.273958394Z     messages: []
2020-01-31T11:35:06.273991847Z     links: [<0.1716.0>]
2020-01-31T11:35:06.273995468Z     dictionary: [{{subscribe,<<"9/33/rpc/#">>},{allow,1580470505433}},
2020-01-31T11:35:06.273998930Z                   {guid,{1580470505435277,221714801755942,4}},
2020-01-31T11:35:06.274001826Z                   {send_pkt,1},
2020-01-31T11:35:06.274004673Z                   {{subscribe,<<"9/33/rw">>},{allow,1580470505432}},
2020-01-31T11:35:06.274007942Z                   {'$logger_metadata$',
2020-01-31T11:35:06.274010787Z                       #{clientid => <<"esp32_D9AB18">>,
2020-01-31T11:35:06.274014005Z                         peername => "49.207.53.3:2896"}},
2020-01-31T11:35:06.274017089Z                   {acl_cache_size,4},
2020-01-31T11:35:06.274019835Z                   {{subscribe,<<"9/33/rpc">>},{allow,1580470505435}},
2020-01-31T11:35:06.274022875Z                   {incoming_pubs,2},
2020-01-31T11:35:06.274025781Z                   {acl_keys_q,
2020-01-31T11:35:06.274028565Z                       {[{subscribe,<<"9/33/rpc">>},
2020-01-31T11:35:06.274031602Z                         {subscribe,<<"9/33/rpc/#">>},
2020-01-31T11:35:06.274034601Z                         {subscribe,<<"9/33/rw">>}],
2020-01-31T11:35:06.274037582Z                        [{publish,<<"9/33/hello">>}]}},
2020-01-31T11:35:06.274040584Z                   {incoming_bytes,318},
2020-01-31T11:35:06.274043325Z                   {outgoing_bytes,4},
2020-01-31T11:35:06.274046096Z                   {rand_seed,
2020-01-31T11:35:06.274048833Z                       {#{bits => 58,jump => #Fun<rand.13.8986388>,
2020-01-31T11:35:06.274051950Z                          next => #Fun<rand.10.8986388>,type => exsss,
2020-01-31T11:35:06.274055029Z                          uniform => #Fun<rand.11.8986388>,
2020-01-31T11:35:06.274068881Z                          uniform_n => #Fun<rand.12.8986388>},
2020-01-31T11:35:06.274072508Z                        [153025356958506172|263484640929313504]}},
2020-01-31T11:35:06.274075365Z                   {recv_msg,2},
2020-01-31T11:35:06.274078108Z                   {recv_pkt,6},
2020-01-31T11:35:06.274080805Z                   {{publish,<<"9/33/hello">>},{allow,1580470505431}}]
2020-01-31T11:35:06.274083779Z     trap_exit: false
2020-01-31T11:35:06.274086482Z     status: running
2020-01-31T11:35:06.274089313Z     heap_size: 2586
2020-01-31T11:35:06.274092022Z     stack_size: 27
2020-01-31T11:35:06.274094757Z     reductions: 22580
2020-01-31T11:35:06.274100469Z   neighbours:
2020-01-31T11:35:06.274106073Z 2020-01-31 11:35:05.436 [error]     supervisor: 'esockd_connection_sup - <0.1716.0>'
2020-01-31T11:35:06.274109166Z     errorContext: connection_crashed
2020-01-31T11:35:06.274111893Z     reason: {badarg,
2020-01-31T11:35:06.274114623Z                 [{ets,lookup_element,
2020-01-31T11:35:06.274117335Z                      [emqx_metrics,'packets.publish.inuse',4],
2020-01-31T11:35:06.274120142Z                      []},
2020-01-31T11:35:06.274122837Z                  {emqx_metrics,update_counter,2,
2020-01-31T11:35:06.274125631Z                      [{file,"/emqx_rel/_checkouts/emqx/src/emqx_metrics.erl"},
2020-01-31T11:35:06.274128593Z                       {line,292}]},
2020-01-31T11:35:06.274131301Z                  {emqx_channel,do_publish,3,
2020-01-31T11:35:06.274148915Z                      [{file,"/emqx_rel/_checkouts/emqx/src/emqx_channel.erl"},
2020-01-31T11:35:06.274152814Z                       {line,413}]},
2020-01-31T11:35:06.274155556Z                  {emqx_connection,with_channel,3,
2020-01-31T11:35:06.274158367Z                      [{file,
2020-01-31T11:35:06.274161081Z                           "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:06.274163958Z                       {line,535}]},
2020-01-31T11:35:06.274166677Z                  {emqx_connection,process_msg,3,
2020-01-31T11:35:06.274169492Z                      [{file,
2020-01-31T11:35:06.274172177Z                           "/emqx_rel/_checkouts/emqx/src/emqx_connection.erl"},
2020-01-31T11:35:06.274175096Z                       {line,282}]},
2020-01-31T11:35:06.274177842Z                  {proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]}
2020-01-31T11:35:06.274180739Z     offender: [{pid,<0.1830.0>},
2020-01-31T11:35:06.274183612Z                {name,connection},
2020-01-31T11:35:06.274186364Z                {mfargs,
2020-01-31T11:35:06.274189036Z                    {emqx_connection,start_link,
2020-01-31T11:35:06.274191789Z                        [[{deflate_options,[]},
2020-01-31T11:35:06.274194518Z                          {max_conn_rate,500},
2020-01-31T11:35:06.274197238Z                          {active_n,100},
2020-01-31T11:35:06.274202105Z                          {zone,external}]]}}]
```

**Comment by turtleDeng** [comment]
Thank you for your participation, please use the stable version 4.0.0 or 4.0.1.

**Comment by daadu** [comment]
@turtleDeng Sure, I will update but can you let me know what the issue is?

**Comment by HJianBo** [comment]
Hi, @daadu The problem is caused that the client sends a repeated `PacketId` of PUBLISH to Broker.

The Broker should drop it, but it causes a process crash.  We have fixed it, it will be released at v4.0.2

**Comment by daadu** [comment]
@HJianBo @turtleDeng @tigercl Thank you for fixing it.

**Comment by daadu** [comment]
Any known reason why the client could be sending repeated `PacketId` of `PUBLISH` ?

**Comment by HJianBo** [comment]
Maybe you should check your client's MQTT library.

It is likely that the client resend the message, or that there is a problem with the client

**Comment by daadu** [comment]
Could be because of client retrying with QoS 2 messages?

**Comment by turtleDeng** [comment]
Thanks for your feedback, we have fixed it in v4.0.2 https://github.com/emqx/emqx/pull/3233

**Comment by daadu** [comment]
I can confirm that the issue is resolved with this version
